### DISCUSSION TITLE
What do you hope will be supported in the next version of Mojo 🔥 0.7? Share your opinion with us

### DISCUSSION THREAD

**Discussion by Eprahim-taha** [modular/modular#1490] [General]
I hope that lifetime will be supported in the next version of Mojo 🔥 What about you? What do you hope will be supported in the next version of Mojo 🔥

**Comment by rarebreed** [comment]
Lifetime/references

I've been holding off on really programming much with the language beyond simple toy exercises, because references will I think, be a pretty drastic change to how you need to think about coding in mojo.

I sincerely hope that mojo won't allow coding with raw `Pointer` types, because honestly, I don't want to deal with segfaults and other memory nightmares.  It will need some way to give access to them, but I don't want willy-nilly access to an unsafe pointer type.  If mojo does allow indiscriminate unsafe programming, I'll probably stick with rust and use some SIMD optimizations hacks along with Godbolt.

I'm actually a little concerned that mojicians will get too comfy programming with raw pointers, so the sooner mojo has a way to pass around references and lifetimes the better.  One reason rust really caught on, despite the learning curve, was that it was ultimately easier to learn, and far safer than C++.

The next thing I want in mojo after lifetimes are Algebraic Data Types.  Traits are relatively simplistic right now without a Sum type.  Also, ADT's will give you the ability to return Option/Maybe types.  While I know mojo has to support Exceptions, it's far more elegant and simpler to have functions that return some kind of Sum (with a Functor-like trait) type.

**Comment by felixfeixu** [comment]
I would like the support of python keyword argument, as tracked by this issue: https://github.com/modularml/mojo/issues/964

Many python functions have many arguments, in python it is not an issue since keyword argument can be used. But when mojo calls python code, it's pain to provide the entire argument list while most of them are optional.

**Reply by trongnamdz1** [reply]
You damn right, still waiting for a fixing up.. smh

**Reply by Brian-M-J** [reply]
> I would like the support of python keyword argument

In [version 24.2](https://docs.modular.com/mojo/changelog#v242-2024-03-28), it's now supported:
> Mojo's Python interoperability now supports passing keyword arguments to Python functions

**Comment by jayzhan211** [comment]
Synchronization primitives.

Python https://docs.python.org/3/library/asyncio-sync.html
Golang https://pkg.go.dev/sync
Rust https://doc.rust-lang.org/std/sync/

**Reply by RichardoZhao** [reply]
I'd like to unsubscribe this project, cause my emailbox has been fed up with all the discussion.
So, I need help, wiill any friend tell me how to achieve that?

**Comment by rarebreed** [comment]
> I'd like to unsubscribe this project, cause my emailbox has been fed up with all the discussion. So, I need help, wiill any friend tell me how to achieve that?

@RichardoZhao Once you are logged into github, go here
https://github.com/watching

Look for modularml/mojo and to the right should be a button "Unwatch".  If you click that, you can either select Ignore, or you can click Custom, and choose which events you will be notifed on (for example, you can deselect Discussions and Issues).

**Comment by Brian-M-J** [comment]
True Windows support without the need for WSL.

**Reply by Eprahim-taha** [reply]
They won't support Windows until they support Intel.  If Mojo now supports Intel, it will run on Windows, but Mojo doesn't support Intel yet

**Reply by soraros** [reply]
@Eprahim-taha What do you mean by Mojo doesn't support Intel? It runs on Intel based system with Linux just fine.

**Reply by Eprahim-taha** [reply]
are you sure

**Reply by felixfeixu** [reply]
> @Eprahim-taha What do you mean by Mojo doesn't support Intel? It runs on Intel based system with Linux just fine.

he means macos, intel version (old mac) is not supported yet, only arm M1 M2 versions are supported

**Reply by soraros** [reply]
@felixf4xu Could be. However, that doesn't explain their doubt("are you sure") for my claim "It runs on Intel based system with Linux just fine." which doesn't have the macOS ambiguity. Nor does it explain their assumption that support for Windows(native) somehow depends on support for macOS(Intel). For instance, I have no problem believing that some work on Linux(Intel) and macOS(ARM) can be reused for macOS(Intel).

**Reply by felixfeixu** [reply]
On Jan 28:
![image](https://github.com/modularml/mojo/assets/84662027/548f01d8-2f9d-4aee-9f86-ecdd1ebd3336)

**Comment by manatlan** [comment]
I'd like a fix on this perf issue : https://github.com/modularml/mojo/issues/1216

Because mojo is still slower than go, rust, nim .. and still slower than java, nodejs, pypy and codon
https://github.com/manatlan/sudoku_resolver/blob/master/RESULTS.md

(With the good old "0.4 mojo version" : it was 10x faster)

**Comment by alfie-nsugh** [comment]
I want support for nvidia gpus

**Reply by rarebreed** [reply]
> I want support for nvidia gpus

I don't know if that will be in the 0.7 release, but according to the Keynote, that will be available once the MAX Engine SDK is released.   Which if I remember correctly, they said should be coming out in first quarter of 2024

**Comment by stefanuytterhoeven** [comment]
Well, concurrency,parallisation is supported. I would like to see examples,examples and examples!
Also some kind of list like 'in python you do this', in mojo it's done like that...

**Reply by alfie-nsugh** [reply]
I completely agree with this. Having some centralised place where you have a bunch of examples would be great. It would greatly speed up learning

**Comment by manatlan** [comment]
I've updated my code&results on my repo : https://github.com/manatlan/sudoku_resolver (differents implems of a sudoku resolver, to compare with others languages)

The spezialized one is quicker ... the simple one is slower ;-(

**Comment by l1t1** [comment]
it works on aarch64 Android+termux+proot-distro Linux
